eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jdbc/JdbcEventStorageEngine.java

```diff
@@ -418,37 +431,40 @@ public TrackingToken createTailToken() {
                 resultSet -> nextAndExtract(resultSet, 1, Long.class),
                 e -> new EventStoreException("Failed to get tail token", e)
         ));
-        return Optional.ofNullable(index)
-                       .map(seq -> GapAwareTrackingToken.newInstance(seq, Collections.emptySet()))
-                       .orElse(null);
+        return createToken(index);
     }
 
     @Override
     public TrackingToken createHeadToken() {
+        return createToken(mostRecentToken());
+    }
+
+    @Override
+    public TrackingToken createTokenAt(Instant dateTime) {
         Long index = transactionManager.fetchInTransaction(() -> executeQuery(
+                getConnection(),
+                connection -> createTokenAt(connection, dateTime),
+                resultSet -> nextAndExtract(resultSet, 1, Long.class),
+                e -> new EventStoreException(format("Failed to get token at [%s]", dateTime), e)
+        ));
+        return index != null ? createToken(index) : createToken(mostRecentToken());
+    }
+
+    private Long mostRecentToken() {
```

Comment:
I feel like the method should include the word `index` somewhere. And since you created this for `head`, we can have something similar for `tail` as we are doing almost the same thing at `createTailToken` - to have it consistent.

Reply:
Renaming to `mostRecentIndex` actually is reasonable, as we're dealing with indices. They might refer to tokens, but it's good to keep that apart at this stage.

Uncertain about a similar private method for the `createTailToken` though. The only reason I introduce the `mostRecentToken`/`mostRecentIndex` method is because I need it to be reused between the `createHeadToken` and `createTokenAt` methods. As we don't have to do a similar thing for tail tokens, I figure this should be fine to be honest.

Let me know what you think!

Reply:
That is completely fine!